### PR TITLE
Adds Rey keeping the past

### PR DESCRIPTION
**Pull request by aruizc**
- Adding Kylo Ren, Killing the Past unit test for non-Heroism/Villany unit test case
- TODO: Uncomment Rey's unit test for Kylo leader once it's implemented

image

### CODE REVIEW

**Comment by AMMayberry1** [comment]
For future PRs, could you please include the card(s) image in the PR description so I can more easily double-check the implementation. Thanks!

File: test/server/cards/02_SHD/units/KyloRenKillingThePast.spec.ts [hunk 1]
```diff
@@ -89,6 +89,21 @@ describe('Kylo Ren, Killing the Past', function() {
                 expect(context.pykeSentinel.getPower()).toBe(2);
                 expect(context.pykeSentinel.getHp()).toBe(3);
             });
+
+            it('gives +2/0 and an Experience token to a non-Villainy/non-Heroism unit', function () {
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/units/KyloRenKillingThePast.spec.ts:93]: Thanks for taking care of this! Do you mind merging this test case into the above one? Can just reset his exhaust status and damage in between attacks and keep this test case name as a comment (see Rukh tests for kind of an example of how we stack multiple on-attack tests into one)

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+                context.setDamage(context.battlefieldMarine, 2);
+                context.player1.clickCard(context.rey);
+                expect(context.player1).toBeAbleToSelectExactly([context.rey, context.battlefieldMarine]);
+                context.player1.clickCard(context.battlefieldMarine);
+                expect(context.battlefieldMarine.getHp()).toBe(3);
+                expect(context.battlefieldMarine).toHaveExactUpgradeNames([]); // no shield
+            });
+
+            // TODO: Uncomment this test after Kylo Ren leader is implemented
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts:80]: It's okay to use unimplemented cards in tests so long as you don't need their ability for anything, so this should work. Also we would want a test with the Kylo unit

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                    }
+                });
+
+                const { context } = contextRef;
+                context.player1.clickCard(context.rey);
+                context.player1.clickCard(context.wildRancor);
+                expect(context.player1).toBeAbleToSelectExactly([context.rey, context.wildRancor]);
+                context.player1.clickCard(context.wildRancor);
+                expect(context.wildRancor.damage).toBe(0);
+                // TODO: Uncomment this line and validate the test
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts:58]: Missed this?

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+                context.setDamage(context.pykeSentinel, 2);
+                context.player1.clickCard(context.rey);
+                expect(context.player1).toBeAbleToSelectExactly([context.rey, context.pykeSentinel]);
+                context.player1.clickCard(context.pykeSentinel);
+                expect(context.pykeSentinel.damage).toBe(0);
+                expect(context.pykeSentinel).toHaveExactUpgradeNames(['shield']);
+            });
+
+            it('heals 2 and gives Shield token to an enemy non-heroism non-villany unit', function () {
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts:41]: Do you mind consolidating these tests into one long one? We try to do that to keep the number lower b/c Jasmine tests are very slow and we have finite github actions time in the free tier (though it's bad unit test practice in general sadly). Can see Rukh tests for kind of an example of how we stack multiple on-attack tests into one, we set up one big describe with all the important pieces.

Similar note for the new Kylo test above

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                });
+
+                const { context } = contextRef;
+
+                context.setDamage(context.battlefieldMarine, 2);
+                context.player1.clickCard(context.rey);
+                expect(context.player1).toBeAbleToSelectExactly([context.rey, context.battlefieldMarine]);
+                context.player1.clickCard(context.battlefieldMarine);
+                expect(context.battlefieldMarine.getHp()).toBe(3);
+                expect(context.battlefieldMarine).toHaveExactUpgradeNames([]); // no shield
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts:77]: Can use `expect(context.battlefieldMarine.isUpgraded()).toBeFalse();` as it's clearer to read

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts (hunk 2 above)
Comment (aruizc) [reply test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts:75]: I included a test with Kylo Ren unit

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts (hunk 3 above)
Comment (aruizc) [reply test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts]: Fixed

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts (hunk 4 above)
Comment (aruizc) [reply test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts]: Refactored

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts (hunk 5 above)
Comment (aruizc) [reply test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts]: Done

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                    }
+                });
+
+                const { context } = contextRef;
+
+                const reset = (passAction = true) => {
+                    context.rey.exhausted = false;
+                    context.rey.damage = 0;
+                    context.wildRancor.damage = 2;
+                    context.pykeSentinel.damage = 0;
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts:40]: Do we need to reset the state for Wild Rancor and Pyke Sentinel? Looks like they're each only involved in one test

File: server/game/cards/02_SHD/units/ReyKeepingThePast.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            condition: (context) => context.source.controller.controlsLeaderOrUnitWithTitle('Kylo Ren')
+        });
+
+        this.addOnAttackAbility({
+            title: 'You may heal 2 damage from a unit. If it’s a non-Heroism unit, give a Shield token to it',
+            targetResolver: {
+                controller: RelativePlayer.Any,
+                cardTypeFilter: WildcardCardType.Unit,
+                immediateEffect: AbilityHelper.immediateEffects.simultaneous([
+                    AbilityHelper.immediateEffects.heal({ amount: 2, optional: true }),
```
[the remark this answers is not in the document]
Comment (aruizc) [reply server/game/cards/02_SHD/units/ReyKeepingThePast.ts]: Done

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts (hunk 6 above)
Comment (aruizc) [reply test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts]: Just for the Wild Rancor as I use it for the attacks, removed the one for pyke sentinel.

File: test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts (hunk 2 above)
Comment (aruizc) [reply test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts:77]: I commented the unit test for the leader as currently we don't have Kylo as a leader yet, once it's implemented I can uncomment and run the test to validate it

**Review by AMMayberry1 — APPROVED** [review]

**Pull request merged**